### DISCUSSION TITLE
`*mut magic_sys::magic_set` cannot be shared between threads safely

### DISCUSSION THREAD

**Discussion by jkoop** [robo9k/rust-magic#180] [Q&A] [answered]
Perhaps I'm doing something stupid, but this code doesn't compile:

```rust
async fn get_content_type(path: &Path) -> Result<String> {
    let cookie = magic::Cookie::open(
        magic::cookie::Flags::ERROR |
        magic::cookie::Flags::MIME_ENCODING |
        magic::cookie::Flags::MIME_TYPE
    )?.load(
        &Default::default()
    )?;  

    Ok(cookie.file(path)?)
}
```

Among other things, it complains:

```plain
error[E0277]: `*mut magic_sys::magic_set` cannot be shared between threads safely
    --> src/server.rs:1620:6
     |  
1620 |     )?;
     |      ^ `*mut magic_sys::magic_set` cannot be shared between threads safely
     |
     = help: within `LoadError<Open>`, the trait `Sync` is not implemented for `*mut magic_sys::magic_set`
     = help: the following other types implement trait `FromResidual<R>`:
               <Result<T, F> as FromResidual<Yeet<E>>>
               <Result<T, F> as FromResidual<Result<Infallible, E>>>

**Comment by robo9k** [comment, marked as answer]
Upon first look, this is a variation of https://github.com/robo9k/rust-magic-sys/issues/28
The problem being that `LoadError<>` contains the original `Cookie<>` so you can recover it, but this is a raw pointer from `libmagic` without enough guarantees to make it `Send` nor `Sync`.

You could map this to your own error and thus discard the private `!Sync` field.
Currently `LoadError` doesn't expose much beyond `std:error::Error`, so if you actually need access to its internals, let me know.

**Comment by jkoop** [comment]
Thanks for the answer. It helped a lot.

For anyone who finds this discussion (including myself), what I did to fix it was I changed the `?` on line 1620 to `.unwrap()`.
